### PR TITLE
Implementation of closure conversion.

### PR DESCRIPTION
**Pull request by rodrigogribeiro**
This PR provides the implementation of the closure conversion algorithm. Additionally, this PR provides bug fixes and syntax sugar for tuples.

### CODE REVIEW

File: runsol.sh
Comment (mbenke) [inline runsol.sh:1]: Why is it empty?

File: src/Solcore/Frontend/Parser/SolcoreParser.y [hunk 1]
```diff
@@ -308,13 +312,14 @@ Pattern : Name PatternList                         {Pat $1 $2}
         | '_'                                      {PWildcard}
         | Literal                                  {PLit $1}
         | '(' Pattern ')'                          {$2}
+        | PatternList                              {Pat (Name "pair") $1}
```
Comment (mbenke) [inline src/Solcore/Frontend/Parser/SolcoreParser.y:315]: Is it always a pair? what about `(x, y, z)` ?

File: test/examples/cases/Compose.solc
Comment (mbenke) [inline test/examples/cases/Compose.solc:1]: Why was main removed?

File: test/examples/cases/NegPair.solc
Comment (mbenke) [inline test/examples/cases/NegPair.solc:1]: Whe remove main, bnot, fromB?

File: src/Solcore/Frontend/Parser/SolcoreParser.y (hunk 1 above)
Comment (rodrigogribeiro) [reply src/Solcore/Frontend/Parser/SolcoreParser.y:315]: I use the name `pair` to denote tuples, not only pairs on the syntax.

File: test/examples/cases/Compose.solc
Comment (rodrigogribeiro) [reply test/examples/cases/Compose.solc:1]: During the debugging, I removed the `main`from this test case. I added the following code:

```
function main () {
   let f = compose(id,id);
  return f(42);
}
```

File: test/examples/cases/NegPair.solc
Comment (rodrigogribeiro) [reply test/examples/cases/NegPair.solc:1]: I removed during debugging. I've restored this file to its old version.

File: runsol.sh
Comment (rodrigogribeiro) [reply runsol.sh:1]: Fixed.

**Pull request merged**